transaction/core/src/signed_contingent_input.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+//! A signed contingent input as described in MCIP #31
+
+use crate::{
+    ring_signature::{
+        CurveScalar, Error as RingSignatureError, GeneratorCache, KeyImage, OutputSecret,
+        PresignedInputRing, RingMLSAG, SignedInputRing,
+    },
+    tx::TxIn,
+    Amount, Commitment, CompressedCommitment, TokenId,
+};
+use alloc::vec::Vec;
+use displaydoc::Display;
+use prost::Message;
+
+/// The "unmasked" data of an amount commitment
+#[derive(Clone, Eq, Message, PartialEq)]
+pub struct UnmaskedAmount {
+    /// The value of the amount commitment
+    #[prost(fixed64, tag = "1")]
+    pub value: u64,
+
+    /// The token id of the amount commitment
+    #[prost(fixed64, tag = "2")]
+    pub token_id: u64,
+
+    /// The blinding factor of the amount commitment
+    #[prost(message, required, tag = "3")]
+    pub blinding: CurveScalar,
+}
+
+/// A signed contingent input is a "transaction fragment" which can be
+/// incorporated into a transaction signed by a counterparty. See MCIP #31 for
+/// motivation.
+#[derive(Clone, Eq, Message, PartialEq)]
+pub struct SignedContingentInput {
+    /// The tx_in which was signed over
+    #[prost(message, required, tag = "1")]
+    pub tx_in: TxIn,
+
+    /// The Ring MLSAG signature, conferring spending authority
+    #[prost(message, required, tag = "2")]
+    pub mlsag: RingMLSAG,
```

Comment:
When a signed contingent input is sent to the consensus service - how would consensus treat it? Does it get included into the actual block structure which can then be scanned?

Reply:
So the idea here is, you can't send this directly to consensus, you can only send a full Transaction to consensus, and we aren't actually changing the consensus service here. So the only thing you can do with this is use it to build a full transaction and then submit that.